[Feature] Implemented a Mute Reason referenced in Issue #385

LCookman opened the pull request:
We have implemented the ability to add a reason when muting a player. The below list of commands are valid command executions, note that this works just like the previous mute, except with a reason. Some drawbacks of the current implementation are below the commands.
```
- /mute <player> [dateDiff] [reason]
- /mute <player> [dateDiff]
- /mute <player> [reason]
- /mute <player>
```
This introduced some new cases to handle. In this version muting with a reason, but no time does not toggle mute like muting without a time, before a reason was an option `[see example 1]`. This is easy to change if you feel that if a player is muted, they should be unmuted if example 1 is run while that player is muted. Also, having an invalid time no longer throws an exception or cancels the command. It simply sets the mute reason to what comes after the name `[see example 2]`. If the user misses a space after the time and includes part of the reason with the time, that portion of the reason will not make it to the actual reason `[see example 3]`, Which will read `Player nightwarrier muted for 9 minutes 59 seconds. Reason: im rico`
```
Examples:
1. /mute delbertina This is a mute
2. /mute CreedTheFreak *d Whoops
3. /mute nightwarrier 10mheyo im rico
```
If you would like to test out this implementation further we have included a compiled Jar file within the zip file. If you have any questions we would be happy to answer them.
[EssentialsX-2.0.1.zip](https://github.com/drtshock/Essentials/files/1309457/EssentialsX-2.0.1.zip)

LCookman committed:
Started work on implementing the mute reason, and decided to write it to the UserData just like the Ban and TempBan reasons

LCookman committed:
Started work on implementing the mute reason, and decided to write it to the UserData just like the Ban and TempBan reasons

LCookman committed:
We have added the ability for the Mute to contain a reason, which is stored for the duration of the mute in the user's data file. Currently we need to add in the mute reason code into some other commands such as /afk and /me and /seen ect. We will also need to fix a problem with the reason when we dont add in a time frame for the mute, which should mute the player indefinatly rather then cancelling the mute and throwing a DataFormat exception.

LCookman committed:
We have added the ability for the Mute to contain a reason, which is stored for the duration of the mute in the user's data file. Currently we need to add in the mute reason code into some other commands such as /afk and /me and /seen ect. We will also need to fix a problem with the reason when we dont add in a time frame for the mute, which should mute the player indefinatly rather then cancelling the mute and throwing a DataFormat exception.

LCookman committed:
Merge pull request #1 from CreedTheFreak/mute/jail_reason_issue#385

Mute reason issue#385

LCookman committed:
fixed some wonky commit thing that happened, thus the following modifications are brought to you by Delbertina.

LCookman committed:
fixed some wonky commit thing that happened, thus the following modifications are brought to you by Delbertina.

LCookman committed:
Merge pull request #3 from CreedTheFreak/update_branch

Modifications Brought To you by Delbertina

delbertina committed:
Added mute format to all msg files

delbertina committed:
Added mute format to all msg files

Commit:
Merge pull request #4 from delbertina/msgs385

Added mute format to all msg files

delbertina committed:
Added check to display mute reason when notifying

delbertina committed:
Added check to display mute reason when notifying

LCookman committed:
Implemented the ability to add a reason while muting a
player. The form of the mute is /mute [Player] [Time] [Reason] or /mute
[Player] [Reason] or /mute [Player] [Time] or /mute [Player]. These are
the various forms of the mute command and acts exactly like the old
command just with a reason.

The final adjustments to the MuteReason

Commit:
Merge pull request #10 from CreedTheFreak/commit_squashing

Final Implementation and the squasing of many commits

LCookman committed:
Update plugin.yml mute command usage

LCookman committed:
Update plugin.yml mute command usage

LCookman committed:
Merge pull request #11 from CreedTheFreak/final_mute_reason_issue#385

Update plugin.yml mute command usage

delbertina commented:
I was curious if this isn't an acceptable way to implement this feature. We could also add a config value or permission to check for. Something along the lines of "allow-mute-reason" as a config value or "essentials.mute.reason" as a permission to use mute reasons.

mdcfe commented on Essentials/src/com/earth2me/essentials/EssentialsPlayerListener.java:
```diff
@@ -80,7 +80,14 @@ public void onPlayerChat(final AsyncPlayerChatEvent event) {
             event.setCancelled(true);
-            user.sendMessage(tl("voiceSilenced"));
+
+            if (user.getMuteReason ().equals ("")) {
```
The space between `getMuteReason` and its brackets doesn't match the style conventions followed by the rest of EssentialsX's code.

mdcfe commented on Essentials/src/com/earth2me/essentials/UserData.java:
```diff
@@ -510,6 +512,30 @@ public void setMuted(boolean set) {
+        }
+    }
+
+    public void setMuteReason (String reason) {
```
The space here is also inconsistent with the rest of EssX.

mdcfe commented on Essentials/src/com/earth2me/essentials/commands/Commandmute.java:
```diff
@@ -66,17 +79,41 @@ public void run(final Server server, final CommandSource sender, final String co
+                        sender.sendMessage(tl("mutedPlayerFor", user.getDisplayName(), muteTime));
+                        user.sendMessage(tl("playerMutedFor", muteTime));
+                    }
+                    else {
```
Another style issue - `else`s should be on the same line as the previous block's closing bracket.

mdcfe commented on Essentials/src/com/earth2me/essentials/commands/Commandafk.java:
```diff
@@ -51,7 +51,12 @@ public void run(Server server, CommandSource sender, String commandLabel, String
         if (message != null && sender != null) {
             if (sender.isMuted()) {
-                throw new Exception(tl("voiceSilenced"));
+                if (sender.getMuteReason ().equals ("")) {
```
I can't help but feel that there is a better way to format these than adding if/else statements everywhere and concatenating the string inside those blocks. Personally, I would:
* Add a `hasMuteReason` method to `UserData` which does `getMuteReason().equals("")`;
* Use a ternary operator when concatenating the strings instead of if/else blocks:
```java
throw new Exception(tl("voiceSilenced") + (sender.hasMuteReason() ?
                tl("muteReason", sender.getMuteReason())));
```

mdcfe commented on Essentials/src/messages.properties:
```diff
@@ -597,3 +598,4 @@ createKitFailed=\u00a74Error occurred whilst creating kit {0}.
 createKitSeparator=\u00a7m-----------------------
 createKitSuccess=\u00a76Created Kit: \u00a7f{0}\n\u00a76Delay: \u00a7f{1}\n\u00a76Link: \u00a7f{2}\n\u00a76Copy contents in the link above into your config.yml.
 whoisUuid=\u00a76 - UUID\:\u00a7r {0}
+muteFormat=\u00a76 Reason: \u00a7c{0}
```
This is a duplicate `muteFormat` key.

mdcfe commented on Essentials/src/messages.properties:
```diff
@@ -494,6 +494,7 @@ vanished=\u00a76You are now completely invisible to normal users, and hidden fro
 versionMismatch=\u00a74Version mismatch\! Please update {0} to the same version.
 versionMismatchAll=\u00a74Version mismatch\! Please update all Essentials jars to the same version.
 voiceSilenced=\u00a76Your voice has been silenced\!
+muteFormat=\u00a74 Reason: {0}
```
Having spaces before the string here means that the string can only be used when concatenating it with another string. In addition, calling it `muteFormat` doesn't seem very consistent with the other uses of `-Format` - `muteReason` would be a more descriptive name for the string.

delbertina committed:
Fixed easy style issues.

delbertina committed:
Fixed easy style issues.

delbertina committed:
Changed muteFormat to muteReason.
Removed space at beginning of muteReason.

delbertina committed:
Changed muteFormat to muteReason.
Removed space at beginning of muteReason.

delbertina committed:
Added method hasMuteReason in UserData.

delbertina committed:
Added method hasMuteReason in UserData.

delbertina committed:
Changed if else blocks to ternary operators. Not sure if way to avoid weird empty string usage.

delbertina committed:
Changed if else blocks to ternary operators. Not sure if way to avoid weird empty string usage.

LCookman committed:
Merge pull request #12 from delbertina/mutes385

Address requested changes to formatting and logic

LCookman committed:
Merge remote-tracking branch 'magnarisa/final_mute_reason_issue#385' into final_mute_reason_issue#385
Updated my local repos for testing purposes

delbertina committed:
Added space when adding reason to string. Dirty :/

delbertina committed:
Added space when adding reason to string. Dirty :/

delbertina committed:
Merge branch 'mutes385' of https://github.com/delbertina/Essentials into mutes385
Updated local copy
Conflicts:
	Essentials/src/com/earth2me/essentials/commands/Commandafk.java
	Essentials/src/com/earth2me/essentials/commands/Commandmail.java
	Essentials/src/com/earth2me/essentials/commands/Commandme.java
	Essentials/src/com/earth2me/essentials/commands/Commandr.java
	Essentials/src/com/earth2me/essentials/commands/Commandseen.java
	Essentials/src/com/earth2me/essentials/commands/Commandwhois.java

delbertina committed:
Added space when adding reason to string.

delbertina committed:
Added space when adding reason to string.

LCookman committed:
Merge branch 'mutes385' of https://github.com/delbertina/Essentials into final_mute_reason_issue#385

delbertina committed:
Added space before reason when a muted player tries to talk.

delbertina committed:
Added space before reason when a muted player tries to talk.

LCookman committed:
Merge branch 'mutes385' of https://github.com/delbertina/Essentials into final_mute_reason_issue#385

delbertina committed:
Fixed logic error in hasMuteReason and errors being thrown from seen and whois.

delbertina committed:
Fixed logic error in hasMuteReason and errors being thrown from seen and whois.

LCookman committed:
Merge branch 'mutes385' of https://github.com/delbertina/Essentials into final_mute_reason_issue#385

LCookman committed:
Merge pull request #14 from delbertina/mutes385

Fixed spacing

delbertina commented:
We have implemented the changes requested, if we missed something please let us know. In taking out the space from the muteReason message, the places that used it required a space to be added. We fixed this, by amending a space before adding the mute reason in the places that needed it. We don’t intend for this change to be accepted, but we don’t know what way of fixing this would best fit the style of this project. The following are some idea we came up.

-Add a message “muteReasonAmend” that is “muteReason” but with a space before to be used when amending onto.
-Similar to how ban reasons are shown, have a “...Reason” version of all the messages that can have a reason in them in the messages.properties files.
-Keep the amended space like how we have it now.

mdcfe commented:
I would suggest the second option, with strings like `voiceSilencedReason` for example, as it would be consistent with other strings in EssentialsX. It also makes formatting more flexible for server owners should they wish to alter it.

delbertina committed:
Added 8 new messages for "...Reason" versions of mute messages and implemented em.

delbertina committed:
Added 8 new messages for "...Reason" versions of mute messages and implemented em.

LCookman committed:
Merge branch 'mutes385' of https://github.com/delbertina/Essentials into final_mute_reason_issue#385
Testing PR

LCookman committed:
Merge branch 'mutes385' of https://github.com/delbertina/Essentials into final_mute_reason_issue#385
Testing PR Changes

delbertina committed:
Added reason stuff to msg command and deleted reference line in seen command. Oops!

delbertina committed:
Added reason stuff to msg command and deleted reference line in seen command. Oops!

LCookman committed:
Merge pull request #15 from delbertina/mutes385

Added 8 new messages for "...Reason" versions of mute messages

LCookman commented:
Alrighties, we think that we have the needed bits and bobs implemented. If not please let us know and we shall make haste to add those in or fix the things. We implemented the second option of adding the various mute reasons within the .properties files.

mdcfe commented on Essentials/src/com/earth2me/essentials/User.java:
```diff
@@ -537,6 +537,7 @@ public boolean checkMuteTimeout(final long currentTime) {
                 setMuteTimeout(0);
                 sendMessage(tl("canTalkAgain"));
                 setMuted(false);
+                setMuteReason ("");
```
Formatting - we don't use spaces between method names and brackets.

mdcfe commented on Essentials/src/com/earth2me/essentials/UserData.java:
```diff
@@ -73,6 +73,7 @@ public final void reloadConfig() {
         godmode = _getGodModeEnabled();
         muted = _getMuted();
         muteTimeout = _getMuteTimeout();
+        muteReason = _getMuteReason ();
```
Formatting again.

mdcfe commented on Essentials/src/com/earth2me/essentials/UserData.java:
```diff
@@ -510,6 +512,34 @@ public void setMuted(boolean set) {
+        if (muteReason != null) {
+            return muteReason;
+        }
+        else {
```
Formatting - `else`s should be on the same line.

mdcfe replied:
Also, why not just return `muteReason` even if it *is* null? We have `hasMuteReason` anyway if someone wants to check whether there is a mute reason in advance, and using `null` has an actual symbolic meaning making it clearer what the returned value means.

LCookman replied:
This is a good point, Initially we had it here since we didn't have `hasMuteReason` but since we now have the ability to check, this is obsolete. I will make the change.

mdcfe commented on Essentials/src/com/earth2me/essentials/UserData.java:
```diff
@@ -510,6 +512,34 @@ public void setMuted(boolean set) {
+    }
+
+    public boolean hasMuteReason(){
+        return !getMuteReason().equals("");
```
Is there any particular reason to call `getMuteReason` instead of just checking the value of `muteReason` directly?

mdcfe commented on Essentials/src/com/earth2me/essentials/commands/Commandmute.java:
```diff
@@ -50,11 +50,24 @@ public void run(final Server server, final CommandSource sender, final String co
+                    muteReason = getFinalArg (args, 1);
+                }
+
+                user.setMuteReason (muteReason);
```
Format.

mdcfe commented on Essentials/src/com/earth2me/essentials/commands/Commandmute.java:
```diff
@@ -50,11 +50,24 @@ public void run(final Server server, final CommandSource sender, final String co
+
             } else {
                 user.setMuted(!user.getMuted());
+                if (!user.getMuted ()) {
```
Format.

mdcfe commented on Essentials/src/com/earth2me/essentials/commands/Commandmute.java:
```diff
@@ -50,11 +50,24 @@ public void run(final Server server, final CommandSource sender, final String co
             } else {
                 user.setMuted(!user.getMuted());
+                if (!user.getMuted ()) {
+                    user.setMuteReason ("");
```
Format.

mdcfe commented on Essentials/src/com/earth2me/essentials/commands/Commandmute.java:
```diff
@@ -66,17 +79,40 @@ public void run(final Server server, final CommandSource sender, final String co
                 if (muteTimestamp > 0) {
-                    sender.sendMessage(tl("mutedPlayerFor", user.getDisplayName(), muteTime));
-                    user.sendMessage(tl("playerMutedFor", muteTime));
+                    if (user.getMuteReason ().equals ("")) {
```
Format.

mdcfe replied:
Also, is there any reason for not using `hasMuteReason`?

mdcfe commented on Essentials/src/com/earth2me/essentials/commands/Commandmute.java:
```diff
@@ -66,17 +79,40 @@ public void run(final Server server, final CommandSource sender, final String co
                 } else {
-                    sender.sendMessage(tl("mutedPlayer", user.getDisplayName()));
-                    user.sendMessage(tl("playerMuted"));
+                    if (user.getMuteReason ().equals ("")) {
```
Format and `hasMuteReason` again.

mdcfe commented on Essentials/src/com/earth2me/essentials/commands/Commandmute.java:
```diff
@@ -66,17 +79,40 @@ public void run(final Server server, final CommandSource sender, final String co
+                        /** Send the player a message, why they were muted **/
+                        user.sendMessage(tl("playerMuted"));
+                    }
+                    else {
```
Format.

mdcfe commented on Essentials/src/com/earth2me/essentials/commands/Commandmute.java:
```diff
@@ -66,17 +79,40 @@ public void run(final Server server, final CommandSource sender, final String co
                 final String message;
                 if (muteTimestamp > 0) {
-                    message = tl("muteNotifyFor", sender.getSender().getName(), user.getName(), muteTime);
+                    if (user.getMuteReason ().equals ("")) {
```
`hasMuteReason` and format.

mdcfe commented on Essentials/src/com/earth2me/essentials/commands/Commandmute.java:
```diff
@@ -66,17 +79,40 @@ public void run(final Server server, final CommandSource sender, final String co
+                        message = tl("muteNotifyFor", sender.getSender().getName(), user.getName(), muteTime);
+                    }
+                    else {
+                        message = (tl("muteNotifyForReason", sender.getSender().getName(), user.getName(), muteTime, user.getMuteReason()));
```
Why is the `tl` call wrapped in brackets?

mdcfe commented on Essentials/src/com/earth2me/essentials/commands/Commandmute.java:
```diff
@@ -66,17 +79,40 @@ public void run(final Server server, final CommandSource sender, final String co
+                    if (user.getMuteReason ().equals ("")) {
+                        message = tl("muteNotifyFor", sender.getSender().getName(), user.getName(), muteTime);
+                    }
+                    else {
```
Format.

mdcfe commented on Essentials/src/com/earth2me/essentials/commands/Commandmute.java:
```diff
@@ -66,17 +79,40 @@ public void run(final Server server, final CommandSource sender, final String co
+                    }
                 } else {
-                    message = tl("muteNotify", sender.getSender().getName(), user.getName());
+                    if (user.getMuteReason ().equals ("")) {
```
Format.

mdcfe commented on Essentials/src/com/earth2me/essentials/commands/Commandmute.java:
```diff
@@ -66,17 +79,40 @@ public void run(final Server server, final CommandSource sender, final String co
+                        message = tl("muteNotify", sender.getSender().getName(), user.getName());
+                    }
+                    else {
+                        message = (tl("muteNotifyReason", sender.getSender().getName(), user.getName(), user.getMuteReason()));
```
Brackets around `tl` again.

mdcfe commented on Essentials/src/com/earth2me/essentials/commands/Commandwhois.java:
```diff
@@ -56,8 +56,9 @@ public void run(final Server server, final CommandSource sender, final String co
-
+        sender.sendMessage(user.hasMuteReason() ? tl("whoisMutedReason", (user.isMuted() ? user.getMuteTimeout() > 0 ?
+                DateUtil.formatDateDiff(user.getMuteTimeout()) : tl("true") : tl("false")),user.getMuteReason()) : tl("whoisMuted", (user.isMuted() ?
+                user.getMuteTimeout() > 0 ? DateUtil.formatDateDiff(user.getMuteTimeout()) : tl("true") : tl("false"))));
```
There is an actual reason to expand these ternary statements into `if ... else`s here, as they are quite complex.

LCookman committed:
Merge remote-tracking branch 'upstream/2.x' into final_mute_reason_issue#385

LCookman committed:
Merge remote-tracking branch 'magnarisa/final_mute_reason_issue#385' into mute_reason_changes
Updated a clean branch with the changes we have made so far.

drtshock commented:
You should also expose this mute reason in /seen https://github.com/EssentialsX/Essentials/blob/2.x/Essentials/src/com/earth2me/essentials/commands/Commandseen.java#L120

drtshock commented:
Also a note, you could join discord to chat about these changes if you'd like. Link is in the readme.

LCookman committed:
Updated the former pull request to include  the specified formatting changes, and small edits to the hasMuteReason and Commandmute dealing with the muteReason. I also included the ability to see the mute reason from the seenOffline method which is identical to the seenOnline code.

LCookman committed:
Updated the former pull request to include  the specified formatting changes, and small edits to the hasMuteReason and Commandmute dealing with the muteReason. I also included the ability to see the mute reason from the seenOffline method which is identical to the seenOnline code.

LCookman committed:
Merge branch 'mute_reason_changes' into final_mute_reason_issue#385
Updating local changes made on mute_reason_changes into the final branch

LCookman commented:
Alright, so I fixed (to my knowledge) all of the formatting issues dealing with the spacing issues, if there are more let me know so I can be sad that I missed them. I have also Updated the instances of checking the muteReason within Commandmute to use `hasMuteReason` which was also updated to check for null instead of an empty string. Finally I added the ability to check if a player is muted while offline by doing /seen while the player is Offline. This is just the same bit of code that is within the seenOnline method that checks if the player has a muteReason. If there is anything else let us know we are happy to keep refining these changes!

LCookman commented:
Just curious, but would you like us to fix those Merge Conflicts with the property files?

mdcfe commented:
@CreedTheFreak If possible, yes. I'd suggest rebasing onto master, pasting the added messages at the bottom then using the bash command at the top of #1734 to sort the messages.

mdcfe commented:
Whoops, didn't mean to close this.

LCookman commented:
Alright thanks! We will get to work on this, when we have some free time.

delbertina committed:
Rebase with sorted messages

LCookman committed:
Merge remote-tracking branch 'delbertina/mutes385test' into final_mute_reason_issue#385
Merged changes into local branch for pushing up to the pull request of
mute reason.

LCookman commented:
Okay, so we were finally able to resolve those conflicts with the rebase and then sorted the files with the given command. Let us know if there is anything else that we need to do. Looks like our commits skyrocketed also, not sure if that was a byproduct of the rebase or not.

drtshock commented on Essentials/src/messages.properties:
```diff
@@ -20,10 +20,10 @@ antiBuildPlace=\u00a74You are not permitted to place\u00a7c {0} \u00a74here.
 antiBuildUse=\u00a74You are not permitted to use\u00a7c {0}\u00a74.
 autoAfkKickReason=You have been kicked for idling more than {0} minutes.
 backAfterDeath=\u00a76Use the /back command to return to your death point.
+backUsageMsg=\u00a76Returning to previous location.
```
You guys are changing a lot of messages here that don't seem relevant to the PR. Can you guys try to minimize the diff in the messages files?

drtshock commented:
It looks like the messages.properties files all have huge diffs on keys that aren't relevant to this PR. If you guys can fix up the diff for those, then I approve :D

mdcfe commented:
Those diffs might be related to #1734 rather than intentional changes here.

drtshock commented:
Probably, but the PR should be updated to reflect that.

delbertina committed:
Sorted messages files with the -f option

delbertina committed:
Fixed error in sorting with number in message name and newlines at end of files.

delbertina committed:
More tweaks. Newlines and ordering switches.

delbertina committed:
One last tweak. New line at the end of the fr file.

mdcfe commented on Essentials/src/com/earth2me/essentials/commands/Commandmute.java:
```diff
@@ -50,11 +50,23 @@ public void run(final Server server, final CommandSource sender, final String co
+                    muteTimestamp = DateUtil.parseDateDiff(time, true);
+                    muteReason = getFinalArg(args, 2);
+                } catch (Exception e) {
+                    user.setMuted((!user.getMuted()));
```
What's the purpose of this line?

LCookman replied:
I don't think there is one now that I think about it. Since regardless of if an exception is caught, we are setting the user to muted after the try catch block. We will remove this and then I believe we also have a fix for the weird property diffs. We ran the command that was given and apparently it sorted them in a different order. If this fix doesn't work we will probably have to do it the long way.

mdcfe commented on Essentials/src/messages.properties:
```diff
@@ -46,8 +46,8 @@ broadcast=\u00a7r\u00a76[\u00a74Broadcast\u00a76]\u00a7a {0}
 buildAlert=\u00a74You are not permitted to build.
 bukkitFormatChanged=Bukkit version format changed. Version not checked.
 burnMsg=\u00a76You set\u00a7c {0} \u00a76on fire for\u00a7c {1} seconds\u00a76.
-cannotStackMob=\u00a74You do not have permission to stack multiple mobs.
```
Your script seems to have different opinions on how to alphabetically sort the keys - if you could minimise the changes to only include the new keys then that would be ideal (perhaps a revert and manually add new keys in the right places).

mdcfe reviewed (changes requested):
Sorry if it seems like we're messing you around a bit, but just a couple of last minute things I noticed. Once that's sorted and the others approve, we can merge.

delbertina committed:
Removed redundant line in mute time try-catch.

delbertina committed:
Merge pull request #17 from delbertina/mutes385test

Sorted messages files with the -f option and other tweaks ...

delbertina commented:
I believe the weird diffs and the redundant line are fixed now. The issue with the sort may have been some OS specific stuff, but it looks all good now to me. I am seeing some weird diffs at the end of the cs, fr, it, zh, and pl message files. Tried a few things to remedy that, but no dice. If that issue is something we need to fix, may be able to dig into things deeper and find a solution. :)

mdcfe reviewed (approved):
Everything looks good to me now, so I'd say we can now squash and merge this. Thanks for the contribution! (and for putting up with our nitpicking :p)

LCookman commented:
Haha no worries, it was kind of our first time as a group working on an open source project. It was fun and we hope to contribute more in the future, without all the simple mistakes :P

mdcfe commented:
@LCookman @delbertina Thanks for the contribution (and apologies for the long delay)! This feature is now available in the [builds 711 and above](https://ci.ender.zone/job/EssentialsX).
